File: analysis/import/SCMLogParser/src/main/java/de/maibornwolff/codecharta/importer/scmlogparser/input/metrics/LinesOfCode.java
package de.maibornwolff.codecharta.importer.scmlogparser.input.metrics;

import de.maibornwolff.codecharta.importer.scmlogparser.input.Modification;

/**
 * this is only an approximation of the correct code churn.
 */
public class LinesOfCode implements Metric {
    private long loc = 0;

    @Override
    public String description() {
        return "Lines Of Code: Approximation for loc. Number of added minus number of deleted lines of all commits for this file.";
    }

    @Override
    public String metricName() {
        return "loc";
    }

    @Override
    public void registerModification(Modification modification) {
        loc += modification.getAdditions();
        loc -= modification.getDeletions();
    }

    /**
     * this is only an approximation of the correct file size.
     * correct only if e.g. --numstat -m --first-parent ist given.
     */
    @Override
    public Number value() {
        return loc >= 0 ? loc : 0;
    }
}

File: analysis/import/SCMLogParser/src/main/kotlin/de/maibornwolff/codecharta/importer/scmlogparser/input/metrics/LinesOfCode.kt
package de.maibornwolff.codecharta.importer.scmlogparser.input.metrics

import de.maibornwolff.codecharta.importer.scmlogparser.input.Modification

/**
 * this is only an approximation of the correct code churn.
 */
class LinesOfCode : Metric {
    private var loc: Long = 0

    override fun description(): String {
        return "Lines Of Code: Approximation for loc. Number of added minus number of deleted lines of all commits for this file."
    }

    override fun metricName(): String {
        return "loc"
    }

    override fun registerModification(modification: Modification) {
        loc += modification.additions
        loc -= modification.deletions
    }

    /**
     * this is only an approximation of the correct file size.
     * correct only if e.g. --numstat -m --first-parent ist given.
     */
    override fun value(): Number {
        return if (loc >= 0) loc else 0
    }
}
